charmcraft/providers.py

```diff
@@ -40,6 +42,56 @@
 }
 
 
+def clean_project_environments(
+    charm_name: str,
+    *,
+    lxd_project: str = "charmcraft",
+    lxd_remote: str = "local",
+) -> List[str]:
+    """Clean up any environments created for project.
+
+    :param charm_name: Name of project.
+    :param lxd_project: Name of LXD project.
+    :param lxd_remote: Name of LXD remote.
+
+    :returns: List of containers deleted.
+    """
+    deleted: List[str] = []
+
+    # Nothing to do if provider is not installed.
+    if not is_provider_available():
+        return deleted
+
+    lxc = lxd.LXC()
+
+    for name in lxc.list_names(project=lxd_project, remote=lxd_remote):
+        match_regex = f"^charmcraft-{charm_name}-.+-.+-.+$"
+        if re.match(match_regex, name):
+            logger.debug("Deleting container: %s", name)
+            lxc.delete(
+                instance_name=name, force=True, project=lxd_project, remote=lxd_remote
+            )
+            deleted.append(name)
+        else:
+            logger.debug("Not deleting container: %s", name)
```

Comment:
Please log with `%r` here and above; getting in one block or the other depends of the regex matching, if we need to debug this for sure we'd want the exact string.

Reply:
Done for both!